### PR TITLE
Mooncake Upgrades

### PR DESCRIPTION
**Pull request by willtebbutt**
In response to a couple of things, I've decided to add much better tools for caching various things in Mooncake. As a result, we can substantially simplify Mooncake's implementation in DifferentiationInterface.

I've tested this locally, so I'm hoping it still works...

### CODE REVIEW

**Comment by willtebbutt** [comment]
(failures because the right version of Mooncake is not yet available, because I messed up doing releases)

**Comment by willtebbutt** [comment]
(codecov is lying -- the line it says is uncovered is definitely covered)

File: DifferentiationInterface/ext/DifferentiationInterfaceMooncakeExt/onearg.jl [hunk 1]
```diff
@@ -1,65 +1,49 @@
...
     f, backend::AutoMooncake, x, ty::NTuple, contexts::Vararg{Context,C}
 ) where {C}
-    y = f(x, map(unwrap, contexts)...)
-    config = get_config(backend)
-    rrule = build_rrule(
-        get_interpreter(),
-        Tuple{typeof(f),typeof(x),typeof.(map(unwrap, contexts))...};
-        debug_mode=config.debug_mode,
-        silence_debug_messages=config.silence_debug_messages,
+    config = backend.config === nothing ? Mooncake.Config() : backend.config
```
Comment (gdalle) [inline DifferentiationInterface/ext/DifferentiationInterfaceMooncakeExt/onearg.jl]:
```suggestion
    config = get_config(backend)
```

File: DifferentiationInterface/ext/DifferentiationInterfaceMooncakeExt/twoarg.jl [hunk 2]
```diff
@@ -1,32 +1,19 @@
...
-    )
-    df! = zero_tangent(f!)
-    y_copy = copy(y)
-    dx_righttype = zero_tangent(x)
-    dy_righttype = zero_tangent(y)
+    target_function = function (f!, y, x, contexts...)
+        f!(y, x, contexts...)
+        return y
+    end
+    cache = prepare_pullback_cache(target_function, f!, y, x, map(unwrap, contexts)...)
```
Comment (gdalle) [inline DifferentiationInterface/ext/DifferentiationInterfaceMooncakeExt/twoarg.jl]: is there a reason why don't use the config here?

File: DifferentiationInterface/ext/DifferentiationInterfaceMooncakeExt/twoarg.jl [hunk 3]
```diff
@@ -1,32 +1,19 @@
...
-        get_interpreter(),
-        Tuple{typeof(f!),typeof(y),typeof(x),typeof.(map(unwrap, contexts))...};
-        debug_mode=config.debug_mode,
-        silence_debug_messages=config.silence_debug_messages,
-    )
-    df! = zero_tangent(f!)
-    y_copy = copy(y)
-    dx_righttype = zero_tangent(x)
-    dy_righttype = zero_tangent(y)
+    target_function = function (f!, y, x, contexts...)
```
Comment (gdalle) [inline DifferentiationInterface/ext/DifferentiationInterfaceMooncakeExt/twoarg.jl:10]: will this new closure slow things down?

**Review by gdalle — COMMENTED** [review]
Thanks! This new cache thingy isn't documented anywhere in Mooncake.jl, is it the new name for rules? Is it an internal or part of the API?

File: DifferentiationInterface/ext/DifferentiationInterfaceMooncakeExt/twoarg.jl (hunk 3 above)
Comment (willtebbutt) [reply DifferentiationInterface/ext/DifferentiationInterfaceMooncakeExt/twoarg.jl:10]: I don't believe so -- Mooncake generally has no problems with closures.

File: DifferentiationInterface/ext/DifferentiationInterfaceMooncakeExt/twoarg.jl (hunk 2 above)
Comment (willtebbutt) [reply DifferentiationInterface/ext/DifferentiationInterfaceMooncakeExt/twoarg.jl]: No, that is a mistake on my part.

Comment (willtebbutt) [reply DifferentiationInterface/ext/DifferentiationInterfaceMooncakeExt/twoarg.jl]: Resolved by latest commit

**Comment by willtebbutt** [comment]
> This new cache thingy isn't documented anywhere in Mooncake.jl, is it the new name for rules?

It's a wrapper around a rule, and all of the various tangents that you need in order to avoid allocations. Unlike `build_rrule`, which just used static information, the `cache` makes use of dynamic information (e.g. array sizes) which it uses to make life easier.

> Is it an internal or part of the API?

It's internal for now. I plan to make it available publicly in short order, but want to make sure it's doing everything as I intend it before telling users about it.

**Pull request merged**